{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.manifold import TSNE\n",
    "\n",
    "from plotly import offline as plotly\n",
    "from plotly import graph_objects\n",
    "\n",
    "from capblood_seq import common as cbs\n",
    "from capblood_seq import viz as cbs_viz\n",
    "from capblood_seq.dataset import Capblood_Seq_Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataset. This downloads it if it doesn't exist already, and loads it into memory\n",
    "dataset = Capblood_Seq_Dataset(pipeline_name=\"RPM_SD\")\n",
    "dataset.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For a combined tSNE, we're only interested in cells that have a single subject or cell type label\n",
    "dataset.filter_unlabeled_cells()\n",
    "dataset.filter_multi_labeled_cells(cbs.SUBJECT_IDS)\n",
    "dataset.filter_multi_labeled_cells(cbs.CELL_TYPES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the combined tSNE, we want to get all cells across all samples and store their\n",
    "# originating sample and subject for plotting\n",
    "\n",
    "# A list of the sample labels for each cell\n",
    "cell_sample_labels = []\n",
    "\n",
    "# A list of the subject labels for each cell\n",
    "cell_subject_labels = []\n",
    "\n",
    "# A matrix containing the gene expression across all samples\n",
    "combined_cell_transcript_counts = numpy.zeros((0,dataset.get_num_genes()))\n",
    "\n",
    "for sample in cbs.SAMPLE_NAMES:\n",
    "    \n",
    "    # For each sample and subject, we get their gene count matrix (normalized)\n",
    "    \n",
    "    for subject_id in cbs.SUBJECT_IDS:\n",
    "        \n",
    "        cell_transcript_counts = dataset.get_transcript_counts(sample, subject_id=subject_id, normalized=True)\n",
    "        \n",
    "        # If this returns None, it means this subject wasn't found in this sample; skipping\n",
    "        if cell_transcript_counts is None:\n",
    "            continue\n",
    "\n",
    "        # Add these cell transcript counts to the combined matrix\n",
    "        combined_cell_transcript_counts = numpy.concatenate(\n",
    "            (combined_cell_transcript_counts, cell_transcript_counts.to_array()))\n",
    "        \n",
    "        # Since we got cells from this sample and subject, we know they all came from it so can just\n",
    "        # extend this list\n",
    "        cell_sample_labels.extend([sample] * cell_transcript_counts.shape[0])\n",
    "        cell_subject_labels.extend([subject_id] * cell_transcript_counts.shape[0])\n",
    "\n",
    "# Convert these into numpy arrays for later masking\n",
    "cell_sample_labels = numpy.array(cell_sample_labels)\n",
    "cell_subject_labels = numpy.array(cell_subject_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transform the normalized gene expression values into PCA space\n",
    "combined_cell_PCs = PCA(n_components=30).fit_transform(\n",
    "    combined_cell_transcript_counts\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transform the combined PCA coordinates into t-SNE space\n",
    "combined_cell_coordinates = TSNE(\n",
    "    verbose=True,\n",
    "    perplexity=30,\n",
    "    n_components=2\n",
    ").fit_transform(\n",
    "    combined_cell_PCs\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What shape to use for each sample\n",
    "\n",
    "SAMPLE_MARKERS = {\n",
    "    \"AM1\": \"triangle-up\",\n",
    "    \"PM1\": \"circle\",\n",
    "    \"AM2\": \"square\",\n",
    "    \"PM2\": \"star\",\n",
    "    \"AM3\": \"cross\",\n",
    "    \"PM3\": \"diamond\"\n",
    "}\n",
    "\n",
    "\n",
    "# Add a different scatter trace for each sample\n",
    "sample_traces = []\n",
    "\n",
    "for sample in cbs.SAMPLE_NAMES:\n",
    "    \n",
    "    # Make a boolean mask representing the cells that belong to this sample\n",
    "    sample_cell_mask = (cell_sample_labels == sample)\n",
    "    \n",
    "    # Grab the sample transformed coordinates\n",
    "    x_values = combined_cell_coordinates[sample_cell_mask, 0]\n",
    "    y_values = combined_cell_coordinates[sample_cell_mask, 1]\n",
    "    \n",
    "    # Grab the associated color for each point in this sample\n",
    "    subject_colors = [cbs_viz.SUBJECT_ID_COLORS[subject_id] \n",
    "                      for subject_id in cell_subject_labels[sample_cell_mask]]\n",
    "\n",
    "    # Make the plotly scatter trace\n",
    "    scatter_trace = graph_objects.Scatter(\n",
    "        x=x_values,\n",
    "        y=y_values,\n",
    "        mode=\"markers\",\n",
    "        marker={\n",
    "            \"size\": 2,\n",
    "            \"color\": subject_colors,\n",
    "            \"symbol\": SAMPLE_MARKERS[sample]\n",
    "        },\n",
    "        name=sample\n",
    "    )\n",
    "    \n",
    "    sample_traces.append(scatter_trace)\n",
    "\n",
    "# Blank the background\n",
    "layout = graph_objects.Layout(\n",
    "    {\n",
    "        \"plot_bgcolor\": \"rgba(255, 255, 255, 0)\",\n",
    "        \"paper_bgcolor\": \"rgba(255, 255, 255, 0)\"\n",
    "    }\n",
    ")\n",
    "\n",
    "# Plot it\n",
    "figure = graph_objects.Figure(data=sample_traces, layout=layout)\n",
    "plotly.iplot(figure)\n",
    "\n",
    "# Save for publication!\n",
    "figure.write_image(os.path.join(\"figures\", \"combined_tSNE_labeled_by_sample.svg\"))\n",
    "figure.write_html(os.path.join(\"figures\", \"combined_tSNE_labeled_by_sample.html\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat for labeling the points by subject\n",
    "\n",
    "# Add a different scatter trace for each sample\n",
    "sample_traces = []\n",
    "\n",
    "for subject_id in cbs.SUBJECT_IDS:\n",
    "    \n",
    "    # Make a boolean mask representing the cells that belong to this subject\n",
    "    subject_id_cell_mask = (cell_subject_labels == subject_id)\n",
    "    \n",
    "    # Grab the sample transformed coordinates\n",
    "    x_values = combined_cell_coordinates[subject_id_cell_mask, 0]\n",
    "    y_values = combined_cell_coordinates[subject_id_cell_mask, 1]\n",
    "    \n",
    "    # Grab the associated marker for each point in this subject\n",
    "    sample_markers = [SAMPLE_MARKERS[sample] for sample in cell_sample_labels[subject_id_cell_mask]]\n",
    "\n",
    "    # Make the plotly scatter trace\n",
    "    scatter_trace = graph_objects.Scatter(\n",
    "        x=x_values,\n",
    "        y=y_values,\n",
    "        mode=\"markers\",\n",
    "        marker={\n",
    "            \"size\": 2,\n",
    "            \"color\": cbs_viz.SUBJECT_ID_COLORS[subject_id],\n",
    "            \"symbol\": sample_markers\n",
    "        },\n",
    "        name=subject_id\n",
    "    )\n",
    "    \n",
    "    sample_traces.append(scatter_trace)\n",
    "\n",
    "# Blank the background\n",
    "layout = graph_objects.Layout(\n",
    "    {\n",
    "        \"plot_bgcolor\": \"rgba(255, 255, 255, 0)\",\n",
    "        \"paper_bgcolor\": \"rgba(255, 255, 255, 0)\"\n",
    "    }\n",
    ")\n",
    "\n",
    "# Plot it\n",
    "figure = graph_objects.Figure(data=sample_traces, layout=layout)\n",
    "plotly.iplot(figure)\n",
    "\n",
    "# Save for publication!\n",
    "figure.write_image(os.path.join(\"figures\", \"combined_tSNE_labeled_by_subject.svg\"))\n",
    "figure.write_html(os.path.join(\"figures\", \"combined_tSNE_labeled_by_subject.html\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
