{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.manifold import TSNE\n",
    "\n",
    "from plotly import offline as plotly\n",
    "from plotly import graph_objects\n",
    "\n",
    "import capblood_seq\n",
    "from capblood_seq import config\n",
    "\n",
    "import pickle\n",
    "\n",
    "SEED=1040"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataset. This downloads it if it doesn't exist already, and loads it into memory\n",
    "dataset = capblood_seq.load_dataset(pipeline_name=\"debris_filtered\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "config.CELL_TYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "config.CELL_SUBTYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the combined tSNE, we want to get all cells across all samples and store their\n",
    "# originating sample and subject for plotting\n",
    "\n",
    "# A list of the sample labels for each cell\n",
    "cell_sample_labels = []\n",
    "\n",
    "# A list of the subject labels for each cell\n",
    "cell_subject_labels = []\n",
    "\n",
    "# A matrix containing the gene expression across all samples\n",
    "combined_transcript_counts = []\n",
    "\n",
    "# A list of the cell types for each cell\n",
    "cell_type_labels = []\n",
    "\n",
    "for sample_index, sample in enumerate(config.SAMPLE_NAMES):\n",
    "    \n",
    "    ged = dataset._sample_datasets[sample]\n",
    "    label_cells = dataset._sample_datasets[sample].get_label_cells()\n",
    "    \n",
    "    cell_transcript_counts = ged.get_cell_transcript_counts()\n",
    "    \n",
    "    for cell_barcode in cell_transcript_counts.row_names:\n",
    "        \n",
    "        subject_id_to_append = None\n",
    "        \n",
    "        for subject_id in config.SUBJECT_IDS:\n",
    "            \n",
    "            if subject_id in label_cells:\n",
    "                \n",
    "                if cell_barcode in label_cells[subject_id]:\n",
    "                    subject_id_to_append = subject_id\n",
    "                    break\n",
    "                    \n",
    "        cell_type_to_append = None\n",
    "                    \n",
    "        for cell_type in config.CELL_TYPES:\n",
    "            \n",
    "            if cell_type in config.CELL_SUBTYPES:\n",
    "                \n",
    "                for cell_subtype in config.CELL_SUBTYPES[cell_type]:\n",
    "                    \n",
    "                    cell_subtype = \"%s %s\" % (cell_subtype, cell_type)\n",
    "                    \n",
    "                    if cell_barcode in label_cells[cell_type]:\n",
    "                        cell_type_to_append = cell_type\n",
    "                        break\n",
    "            else:\n",
    "                if cell_barcode in label_cells[cell_type]:\n",
    "                    cell_type_to_append = cell_type\n",
    "                    break\n",
    "        \n",
    "        cell_subject_labels.append(subject_id_to_append)\n",
    "        cell_type_labels.append(cell_type_to_append)\n",
    "        \n",
    "    cell_transcript_counts = cell_transcript_counts.to_array()\n",
    "    \n",
    "    # Since we got cells from this sample and subject, we know they all came from it so can just\n",
    "    # extend this list\n",
    "    cell_sample_labels.extend([sample] * cell_transcript_counts.shape[0])\n",
    "    \n",
    "    # Add these cell transcript counts to the combined matrix\n",
    "    combined_transcript_counts.append(cell_transcript_counts)\n",
    "\n",
    "# Convert these into numpy arrays for later masking\n",
    "cell_sample_labels = numpy.array(cell_sample_labels)\n",
    "cell_subject_labels = numpy.array(cell_subject_labels)\n",
    "cell_type_labels = numpy.array(cell_type_labels)\n",
    "\n",
    "combined_transcript_counts = numpy.concatenate(\n",
    "   combined_transcript_counts, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initiliaze variationan autoencoder and training parameters\n",
    "n_epochs = 50\n",
    "learning_rate = 1e-3\n",
    "# Save the training weights\n",
    "latent_pickle_file_name = os.path.join(\"data\", \"Dobreva2020\", \"dobreva2020_nepoch_%i_lr_%.1e_latent.pickle\" % (n_epochs, learning_rate))\n",
    "\n",
    "# Transform the normalized gene expression values into PCA space\n",
    "with open(latent_pickle_file_name, 'rb') as latent_pickle_file:\n",
    "    latent = pickle.load(latent_pickle_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Transform the combined PCA coordinates into t-SNE space\n",
    "combined_cell_coordinates = TSNE(\n",
    "    verbose=True,\n",
    "    perplexity=30,\n",
    "    n_components=2,\n",
    "    random_state=SEED\n",
    ").fit_transform(\n",
    "    latent\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "config.SUBJECT_ID_COLORS[None] = \"Grey\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What shape to use for each sample\n",
    "\n",
    "SAMPLE_MARKERS = {\n",
    "    \"AM1\": \"triangle-up\",\n",
    "    \"PM1\": \"circle\",\n",
    "    \"AM2\": \"square\",\n",
    "    \"PM2\": \"star\",\n",
    "    \"AM3\": \"cross\",\n",
    "    \"PM3\": \"diamond\"\n",
    "}\n",
    "\n",
    "\n",
    "# Add a different scatter trace for each sample\n",
    "sample_traces = []\n",
    "\n",
    "for sample in config.SAMPLE_NAMES:\n",
    "    \n",
    "    # Make a boolean mask representing the cells that belong to this sample\n",
    "    sample_cell_mask = (cell_sample_labels == sample) & (cell_type_labels != None)\n",
    "    \n",
    "    # Grab the sample transformed coordinates\n",
    "    x_values = combined_cell_coordinates[sample_cell_mask, 0]\n",
    "    y_values = combined_cell_coordinates[sample_cell_mask, 1]\n",
    "    \n",
    "    # Grab the associated color for each point in this sample\n",
    "    subject_colors = [config.SUBJECT_ID_COLORS[subject_id] \n",
    "                      for subject_id in cell_subject_labels[sample_cell_mask]]\n",
    "\n",
    "    # Make the plotly scatter trace\n",
    "    scatter_trace = graph_objects.Scatter(\n",
    "        x=x_values,\n",
    "        y=y_values,\n",
    "        mode=\"markers\",\n",
    "        marker={\n",
    "            \"size\": 3,\n",
    "            \"color\": subject_colors,\n",
    "            \"symbol\": SAMPLE_MARKERS[sample]\n",
    "        },\n",
    "        name=sample\n",
    "    )\n",
    "    \n",
    "    sample_traces.append(scatter_trace)\n",
    "\n",
    "# Blank the background\n",
    "layout = graph_objects.Layout(\n",
    "    {\n",
    "        \"plot_bgcolor\": \"rgba(255, 255, 255, 0)\",\n",
    "        \"paper_bgcolor\": \"rgba(255, 255, 255, 0)\"\n",
    "    }\n",
    ")\n",
    "\n",
    "# Plot it\n",
    "figure = graph_objects.Figure(data=sample_traces, layout=layout)\n",
    "plotly.iplot(figure)\n",
    "\n",
    "# Save for publication!\n",
    "figure.write_image(os.path.join(\"figures\", \"combined_tSNE_labeled_by_sample.svg\"))\n",
    "figure.write_html(os.path.join(\"figures\", \"combined_tSNE_labeled_by_sample.html\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat for labeling the points by subject\n",
    "\n",
    "# Add a different scatter trace for each sample\n",
    "sample_traces = []\n",
    "\n",
    "for subject_id in config.SUBJECT_IDS:\n",
    "    \n",
    "    # Make a boolean mask representing the cells that belong to this subject\n",
    "    subject_id_cell_mask = (cell_subject_labels == subject_id) & (cell_type_labels != None)\n",
    "    \n",
    "    # Grab the sample transformed coordinates\n",
    "    x_values = combined_cell_coordinates[subject_id_cell_mask, 0]\n",
    "    y_values = combined_cell_coordinates[subject_id_cell_mask, 1]\n",
    "    \n",
    "    # Grab the associated marker for each point in this subject\n",
    "    sample_markers = [SAMPLE_MARKERS[sample] for sample in cell_sample_labels[subject_id_cell_mask]]\n",
    "\n",
    "    # Make the plotly scatter trace\n",
    "    scatter_trace = graph_objects.Scatter(\n",
    "        x=x_values,\n",
    "        y=y_values,\n",
    "        mode=\"markers\",\n",
    "        marker={\n",
    "            \"size\": 2.5,\n",
    "            \"color\": config.SUBJECT_ID_COLORS[subject_id]\n",
    "        },\n",
    "        name=subject_id\n",
    "    )\n",
    "    \n",
    "    sample_traces.append(scatter_trace)\n",
    "\n",
    "# Blank the background\n",
    "layout = graph_objects.Layout(\n",
    "    {\n",
    "        \"plot_bgcolor\": \"rgba(255, 255, 255, 0)\",\n",
    "        \"paper_bgcolor\": \"rgba(255, 255, 255, 0)\"\n",
    "    }\n",
    ")\n",
    "\n",
    "# Plot it\n",
    "figure = graph_objects.Figure(data=sample_traces, layout=layout)\n",
    "plotly.iplot(figure)\n",
    "\n",
    "# Save for publication!\n",
    "figure.write_image(os.path.join(\"figures\", \"combined_tSNE_labeled_by_subject.svg\"))\n",
    "figure.write_html(os.path.join(\"figures\", \"combined_tSNE_labeled_by_subject.html\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
