{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import capblood_seq\n",
    "import scvi\n",
    "from scvi.dataset import GeneExpressionDataset\n",
    "import numpy\n",
    "from sparsedat import wrappers \n",
    "import scipy\n",
    "from scvi.models.vae import VAE\n",
    "from scvi.inference import UnsupervisedTrainer\n",
    "from scvi.models.scanvi import SCANVI\n",
    "import scanpy as sc\n",
    "from plotly import offline as plotly\n",
    "from sklearn.manifold import TSNE\n",
    "from plotly import graph_objects\n",
    "from umap import UMAP\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "from sklearn.cluster import KMeans\n",
    "import torch\n",
    "import pickle\n",
    "import os\n",
    "import random\n",
    "import pandas as pd\n",
    "from scvi import set_seed\n",
    "import pickle\n",
    "from scrapi.dataset import Gene_Expression_Dataset as GED\n",
    "from capblood_seq import config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SEED=1040"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "capblood_seq_data = capblood_seq.load_dataset(\"data\", pipeline_name=\"debris_filtered\")\n",
    "# Combine transcript counts from the different samples into one big matrix\n",
    "\n",
    "cell_sample_index = []\n",
    "combined_transcript_counts = []\n",
    "\n",
    "for sample_index,sample in enumerate(config.SAMPLE_NAMES):\n",
    "    ged = capblood_seq_data._sample_datasets[sample]\n",
    "    cell_transcript_counts = ged.get_cell_transcript_counts()\n",
    "    cell_transcript_counts = cell_transcript_counts.to_array()\n",
    "    combined_transcript_counts.append(cell_transcript_counts)\n",
    "    cell_sample_index.extend([sample_index]*cell_transcript_counts.shape[0])\n",
    "\n",
    "combined_transcript_counts = numpy.concatenate(\n",
    "   combined_transcript_counts, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Populate the gene expression set with data\n",
    "ged = GeneExpressionDataset()\n",
    "\n",
    "ged.populate_from_data(\n",
    "    combined_transcript_counts,\n",
    "    gene_names=capblood_seq_data.gene_list,\n",
    "    batch_indices=cell_sample_index\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initiliaze variationan autoencoder and training parameters\n",
    "n_epochs = 50\n",
    "learning_rate = 1e-3\n",
    "# Save the training weights\n",
    "latent_pickle_file_name = os.path.join(\"data\", \"Dobreva2020\", \"dobreva2020_nepoch_%i_lr_%.1e_latent.pickle\" % (n_epochs, learning_rate))\n",
    "weights_pickle_file_name = os.path.join(\"data\", \"Dobreva2020\", \"dobreva2020_nepoch_%i_lr_%.1e_weights.pickle\" % (n_epochs, learning_rate))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vae = VAE(ged.nb_genes,n_batch=ged.n_batches)\n",
    "trainer=UnsupervisedTrainer(vae,ged,train_size=0.8,frequency=1,seed=SEED)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you can't load existing latent space, train!\n",
    "if not os.path.exists(latent_pickle_file_name):\n",
    "\n",
    "    set_seed(SEED)\n",
    "    \n",
    "    trainer.train(n_epochs=n_epochs, lr=learning_rate)\n",
    "    torch.save(trainer.model.state_dict(), weights_pickle_file_name)\n",
    "    \n",
    "    full = trainer.create_posterior(trainer.model, ged, indices=numpy.arange(len(ged)))\n",
    "    latent, _, _ = full.sequential().get_latent()\n",
    "    \n",
    "    with open(latent_pickle_file_name, 'wb') as latent_pickle_file:\n",
    "        pickle.dump(latent, latent_pickle_file, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "\n",
    "else:\n",
    "    \n",
    "    weights_pickle_file = torch.load(weights_pickle_file_name)\n",
    "    trainer.model.load_state_dict(weights_pickle_file)\n",
    "    \n",
    "    with open(latent_pickle_file_name, 'rb') as latent_pickle_file:\n",
    "        latent = pickle.load(latent_pickle_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tsne = TSNE(n_components = 2,random_state=SEED).fit_transform(latent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters = AgglomerativeClustering(n_clusters=13).fit_predict(latent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "traces = []\n",
    "\n",
    "for cluster_index in range(clusters.max()+1):\n",
    "    \n",
    "    x = tsne[clusters == cluster_index, 0]\n",
    "    y = tsne[clusters == cluster_index, 1]\n",
    "    \n",
    "    trace = graph_objects.Scatter(\n",
    "        x=x,\n",
    "        y=y,\n",
    "        name=\"Cluster %i\" % cluster_index,\n",
    "        mode=\"markers\"\n",
    "    )\n",
    "    \n",
    "    traces.append(trace)\n",
    "\n",
    "figure = graph_objects.Figure(traces)\n",
    "\n",
    "plotly.iplot(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_cell_marker_map = {}\n",
    "cluster_cell_marker_map[3] = 'CD4 T Cells'\n",
    "cluster_cell_marker_map[7] = 'CD4 T Cells'\n",
    "\n",
    "cluster_cell_marker_map[2] = 'CD8 T Cells'\n",
    "cluster_cell_marker_map[9] = 'CD8 T Cells'\n",
    "\n",
    "cluster_cell_marker_map[4] = 'NK Cells'\n",
    "cluster_cell_marker_map[5] = 'NK Cells'\n",
    "\n",
    "cluster_cell_marker_map[11] = 'B Cells'\n",
    "cluster_cell_marker_map[12] = 'B Cells'\n",
    "\n",
    "cluster_cell_marker_map[0] = 'CD14 Monocytes'\n",
    "cluster_cell_marker_map[10] = 'CD14 Monocytes'\n",
    "\n",
    "cluster_cell_marker_map[6] = 'CD16 Monocytes'\n",
    "\n",
    "cluster_cell_marker_map[8] = 'Dendritic Cells'\n",
    "\n",
    "\n",
    "cluster_cell_marker_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_index = 0\n",
    "\n",
    "for sample_index, sample in enumerate(config.SAMPLE_NAMES):\n",
    "    \n",
    "    ged = capblood_seq_data._sample_datasets[sample]\n",
    "    \n",
    "    num_cells = ged.num_cells\n",
    "    \n",
    "    # Rows are cells\n",
    "    cell_barcodes = numpy.array(ged._cell_transcript_counts.row_names)\n",
    "    \n",
    "    for cluster, label in cluster_cell_marker_map.items():\n",
    "        \n",
    "        cluster_mask = clusters == cluster\n",
    "        cluster_mask = cluster_mask[cell_index:cell_index+num_cells]\n",
    "        \n",
    "        label_barcodes = cell_barcodes[cluster_mask]\n",
    "        label = cluster_cell_marker_map[cluster]\n",
    "        \n",
    "        ged.label_cells(label, label_barcodes)\n",
    "        \n",
    "    ged.save_labels()\n",
    "    \n",
    "    cell_index += num_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENE = \"COX5A\"\n",
    "\n",
    "gene_index = capblood_seq_data.gene_list.index(GENE)\n",
    "\n",
    "traces = []\n",
    "    \n",
    "x = tsne[:, 0]\n",
    "y = tsne[:, 1]\n",
    "\n",
    "trace = graph_objects.Scatter(\n",
    "    x=x,\n",
    "    y=y,\n",
    "    name=\"Cluster %i\" % cluster_index,\n",
    "    mode=\"markers\",\n",
    "    marker={\n",
    "        \"color\": combined_transcript_counts[:, gene_index].flatten()\n",
    "    }\n",
    ")\n",
    "\n",
    "traces.append(trace)\n",
    "\n",
    "figure = graph_objects.Figure(traces)\n",
    "\n",
    "plotly.iplot(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_mask(index_array,num_elements):\n",
    "    mask_array = numpy.zeros(num_elements, dtype=int)\n",
    "    mask_array[index_array] = 1\n",
    "    return mask_array\n",
    "\n",
    "cell_types = ['B Cells','CD4 T Cells','CD8 T Cells','Dendritic Cells','CD14 Monocytes','CD16 Monocytes','Neutrophils','NK Cells','T Cells','Monocytes']\n",
    "lee_cell_stat_pd = pd.DataFrame(columns=cell_types)\n",
    "lee_cell_stat_pd[\"Subject\"] = ['1','2','3','4','5','6','7','8','9','10','11','12','13','14','15','16','17','18','19','20','21','22','23','24']\n",
    "lee_cell_stat_pd = lee_cell_stat_pd.set_index(\"Subject\")\n",
    "for col in lee_cell_stat_pd.columns:\n",
    "    lee_cell_stat_pd[col].values[:] = 0\n",
    "\n",
    "combined_batch_indices = cell_sample_index\n",
    "cell_index_offset = 0\n",
    "for cluster in range(clusters.max() + 1):\n",
    "    subject_batch_counter = 0\n",
    "    cluster_mask = clusters == cluster\n",
    "    num_cells = cluster_mask.sum()\n",
    "\n",
    "    cell_index_offset = 0\n",
    "    \n",
    "    for batch_index in range(max(combined_batch_indices)+1):\n",
    "\n",
    "        sample = test_samples[batch_index]\n",
    "\n",
    "        ged = capblood_seq_data._sample_datasets[sample]\n",
    "\n",
    "        batch_mask = numpy.array(combined_batch_indices) == batch_index\n",
    "        batch_cell_mask = batch_mask & cluster_mask\n",
    "        total_batch_num_cells = batch_mask.sum()\n",
    "        num_cells_cluster = batch_cell_mask.sum()\n",
    "\n",
    "        for subject_id in capblood_seq.config.SUBJECT_IDS:\n",
    "            \n",
    "            try:\n",
    "                subject_cell_barcodes = ged.get_cells(subject_id)\n",
    "            except:\n",
    "                subject_batch_counter += 1\n",
    "                continue\n",
    "\n",
    "            subject_sample_cell_indices = [ged._cell_transcript_counts.row_names.index(cell_barcode) for cell_barcode in subject_cell_barcodes]\n",
    "\n",
    "            subject_cell_indices = [x + cell_index_offset for x in subject_sample_cell_indices]\n",
    "\n",
    "            subject_batch_mask = create_mask(subject_cell_indices,len(batch_mask))\n",
    "\n",
    "\n",
    "            # number of cells in this batch for this person\n",
    "            num_subject_batch_mask = subject_batch_mask.sum()\n",
    "\n",
    "            #Get mask for subject,cluster\n",
    "            subject_batch_cluster_mask = batch_cell_mask & subject_batch_mask\n",
    "\n",
    "            #number of cells in this batch,this subject\n",
    "            num_subject_batch_cluster_mask = subject_batch_cluster_mask.sum()\n",
    "\n",
    "            percentage_of_cells = num_subject_batch_cluster_mask*100/num_subject_batch_mask\n",
    "\n",
    "            print(\"Cluster: \")\n",
    "            print(cluster_cell_marker_map[cluster])\n",
    "            print(subject_batch_counter)\n",
    "\n",
    "            if 'B Cells' in cluster_cell_marker_map[cluster]:\n",
    "                lee_cell_stat_pd['B Cells'].iloc[subject_batch_counter] += percentage_of_cells\n",
    "            if 'CD4 T Cells' in cluster_cell_marker_map[cluster]:\n",
    "                lee_cell_stat_pd['CD4 T Cells'].iloc[subject_batch_counter] += percentage_of_cells\n",
    "            if 'CD8 T Cells' in cluster_cell_marker_map[cluster]:\n",
    "                lee_cell_stat_pd['CD8 T Cells'].iloc[subject_batch_counter] += percentage_of_cells\n",
    "            if 'Dendritic Cells' in cluster_cell_marker_map[cluster]:\n",
    "                lee_cell_stat_pd['Dendritic Cells'].iloc[subject_batch_counter] += percentage_of_cells\n",
    "            if 'CD14 Monocytes' in cluster_cell_marker_map[cluster]:\n",
    "                lee_cell_stat_pd['CD14 Monocytes'].iloc[subject_batch_counter] += percentage_of_cells\n",
    "            if 'CD16 Monocytes' in cluster_cell_marker_map[cluster]:\n",
    "                lee_cell_stat_pd['CD16 Monocytes'].iloc[subject_batch_counter] += percentage_of_cells\n",
    "            if 'Neutrophils' in cluster_cell_marker_map[cluster]:\n",
    "                lee_cell_stat_pd['Neutrophils'].iloc[subject_batch_counter] += percentage_of_cells\n",
    "            if 'NK Cells' in cluster_cell_marker_map[cluster]:\n",
    "                lee_cell_stat_pd['NK Cells'].iloc[subject_batch_counter] += percentage_of_cells\n",
    "            if ('Monocytes' or 'Dendritic Cells') in cluster_cell_marker_map[cluster]:\n",
    "                lee_cell_stat_pd['Monocytes'].iloc[subject_batch_counter] += percentage_of_cells\n",
    "            if 'T Cells' in cluster_cell_marker_map[cluster]:\n",
    "                lee_cell_stat_pd['T Cells'].iloc[subject_batch_counter] += percentage_of_cells\n",
    "\n",
    "            subject_batch_counter += 1\n",
    "\n",
    "    \n",
    "        cell_index_offset += ged.num_cells\n",
    "            \n",
    "lee_cell_stat_pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lee_cell_stat_pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
